### PR TITLE
Fix 5706 Block Positioning when nesting and blocks span columns

### PR DESCRIPTION
**Pull request by NealGooch**
## :bookmark_tabs: Summary

Block diagrams positioning goes wrong when nesting blocks and blocks span columns

Resolves #6707

## :straight_ruler: Design Decisions

Fix

### :clipboard: Tasks

Make sure you

- [ ] :book: have read the [contribution guidelines](https://mermaid.js.org/community/contributing.html)
- [ ] :computer: have added necessary unit/e2e tests.
- [ ] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://mermaid.js.org/community/contributing.html#update-documentation) is used for all new features.
- [ ] :butterfly: If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Review by knsv — COMMENTED** [review]
Hey @NealGooch — first of all, sorry for the long wait on this review. This PR has been sitting for over a year and that's not the experience we want for contributors. Let's get this across the finish line.

## What's working well

🎉 **[praise]** Great bug catch! The fix is surgically precise — `block.widthInColumns` → `child.widthInColumns` on line 64 of `layout.ts` is clearly correct. When normalizing a child's rendered width to find the per-column unit width, you need to divide by that *child's* column span, not the parent's. Using the parent's span was causing completely wrong size calculations for nested blocks with multi-column children, which explains the overlapping and mispositioned elements in issue #5706.

🎉 **[praise]** One-line fix, right at the root cause. This is the kind of change that's easy to get right and hard to argue with.

## Things to address

🟡 **[important]** **Missing E2E visual regression test.** Block diagram layout changes must have a Cypress snapshot test to prevent future regressions. Please add a test case in `cypress/integration/rendering/block.spec.js` (or similar) using `imgSnapshotTest()` with the reproduction case from issue #5706:

```
block-beta
    columns 4

    block:0_0:4
        columns 6
            Example:2
        space:2
        ExampleOther:2
        ExampleOther:1
        block:0_0_0:6
            a b c d e f g h
        end
    end
```

This is especially important because this is shared layout code — a future refactor could easily re-introduce the same bug without visual coverage.

🟡 **[important]** **Missing unit test.** It would be great to add a unit test for `getMaxChildSize()` that covers multi-column children. The existing test file (`layout.spec.ts`) only tests `calculateBlockPosition`. A test that constructs a mock `Block` with children having different `widthInColumns` values and verifies the returned max width would nail this down. This provides faster feedback than E2E alone.

🟡 **[important]** **Missing changeset.** Since this is a bug fix, please generate one:
```bash
pnpm changeset
```
Select `packages/mermaid`, `patch` bump, and prefix the description with `fix:` (e.g., `fix: correct block positioning when nested blocks span multiple columns`).

## Self-check

- [x] At least one 🎉 praise item
- [x] No duplicate comments
- [x] Severity tally: 0 🔴 / 3 🟡 / 0 🟢 / 0 💡 / 2 🎉
- [x] Verdict: Approved
- [x] Tone check: collaborative and appreciative

Thanks for the contribution — the fix itself is spot-on. Once the test and changeset are added, this should be ready to go. Let me know if you need any help with the Cypress test setup!

**Comment by knsv** [comment]
Here's why that breaks with the PR fix:

  1. Height divergence after resizing. In setBlockSizes, step 3 (lines 110–112) initially makes all root children the same height (maxHeight). But step 4 (lines 121–123) calls
  setBlockSizes recursively again, and at line 186–191, each block's height gets rewritten to match its internal layout:

  - group1 (2 columns, 4 children → 2 rows internally) → tall
  - group2 (auto columns, 7 children → 1 row internally) → short

  2. The y-position formula breaks. When positioning group2 at py=2:

  group2.y = parent.y - parent.h/2 + 2 * (group2.height + padding) + group2.height/2 + padding

  It computes the offset for rows 0 and 1 using group2.height — but row 1 actually contains group1, which is taller than group2. The formula underestimates the space taken by row
   1, so group2 gets placed too high, overlapping group1.

  3. Why wasn't this visible before the fix? With the original code (block.widthInColumns — the parent's span), getMaxChildSize returns an inflated maxWidth. This causes the "too
   small sibling" branch (line 142–163) to trigger more aggressively during the second setBlockSizes pass, forcing:

  height = siblingHeight;  // line 147 — forces uniform height

  This masks the y-positioning bug by keeping all children at the same height. The PR fix produces correct per-column widths, so children keep their natural heights — and the
  pre-existing row-height bug surfaces.

  In short: the fix is correct for width calculation, but it exposes a latent bug in layoutBlocks where the y-positioning formula assumes uniform row heights. The real fix needs
  to also track per-row max heights and use cumulative row offsets instead of py * (child.height + padding).

**Review by knsv — APPROVED** [review]
Ready to approve this one now!
